belief_manager.py

```diff
@@ -0,0 +1,578 @@
+      value is a tuple of the X and Y position of the transmitter, or, if
+      non-associated readings were provided, a simple list of these tuples. """
+    if type(readings) == dict:
+      positions = {}
+      for transmitter, reading in readings.iteritems():
+        positions[transmitter] = \
+            self._single_point_strength_distance(self._filter.position(),
+                                                 reading[0], reading[1])
+    else:
+      positions = []
+      for reading in readings:
+        location = self._single_point_strength_distance(self._filter.position(),
+                                                        reading[0], reading[1])
+        positions.append(location)
+
+    return positions
+
+  def _distance_from_past_states(self, transmitters):
+    """ Uses data from previous states as a more accurate method for calculating
+    transmitter positions. The way this works is that the drone gets two LOBs on
+    the transmitter at two different times. Where those LOBs cross is the
+    location of the transmitter.
+    transmitters: The indexes of the transmitters in the state to use.
+    Returns: A dict of tuples, with each tuple containing the estimated x and y
+      position of the transmitter. The keys are the indices of the corresponding
+      transmitters in the state. """
+    logger.debug("Past lobs: Calculating distance.")
+    positions = {}
+    for transmitter in transmitters:
+      # Check whether we have enough data to use the filtered state.
+      current_lob = self._filter.state()[transmitter]
+      current_position = self._filter.position()
+
+      # To find out where they cross, we first have to translate the lobs into
+      # actual lines. Start with the current one.
+      intersection = self._calculate_intersection(transmitter - 4, current_lob,
+                                                  current_position)
+
+      positions[transmitter] = intersection
+
+    return positions
+
+  def _calculate_intersection(self, transmitter, new_lob, position):
+    """ Calculates where two LOBs on the same transmitter intersect.
+    The new LOB is supplied, and the old LOB is taken from the saved state
+    information.
+    Args:
+      transmitter: The index in the _past_lobs array of the transmitter in
+      question.
+      new_lob: The new LOB to this transmitter.
+      position: The current position of the drone to use in the calculation.
+    Returns:
+      A tuple containing an x and y coordinate for the point of intersection.
+    """
+    # To find out where they cross, we first have to translate the lobs into
+    # actual lines. Start with the current one.
+    current_slope = np.tan(new_lob)
+    current_intercept = position[self._Y] - current_slope * \
+                        position[self._X]
+
+    # Translate the old one.
+    past_data = self._past_lobs[transmitter]
+    past_lob = past_data[0]
```

Comment:
There is a comment somewhere that explains exactly how _past_lobs is organized. I have, however, extended the comment in this immediate area in order to make it completely clear.